src_lockfree/bounded_stack.ml

```diff
@@ -0,0 +1,156 @@
+open Picos
+
+type 'a node = Nil | Cons of { value : 'a; tail : 'a node; capacity : int }
+
+type 'a t = {
+  head : 'a node Atomic.t;
+  cons_waiters : Trigger.t list Atomic.t;
+  prod_waiters : Trigger.t list Atomic.t;
+  capacity : int;
+}
+
+let create ?(capacity = Int.max_int) () =
+  let head = Atomic.make_contended Nil in
+  let cons_waiters = Atomic.make_contended [] in
+  let prod_waiters = Atomic.make_contended [] in
+  { head; cons_waiters; prod_waiters; capacity = max capacity 1 }
+  |> Multicore_magic.copy_as_padded
+
+let is_empty t = Atomic.get t.head == Nil
+
+let rec signal_all waiters =
+  let triggers = Atomic.get waiters in
+  if triggers != [] then
+    if Atomic.compare_and_set waiters triggers [] then
+      List.iter Trigger.signal triggers
```

Comment:
Note that this signals the triggers in LIFO order.  Ideally we might want to signal triggers in FIFO order, to hopefully improve fairness, and try to avoid signaling all of them, to avoid the [thundering herd](https://en.wikipedia.org/wiki/Thundering_herd_problem) phenomena.  At this point I mention these mostly as things to work on later.